crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -660,9 +660,9 @@ where
                 debug_assert!(self.current_assignment.is_none());
                 self.current_assignment = Some(node.into());
                 // TODO walrus in comprehensions is implicitly nonlocal
+                self.visit_expr(&node.value);
```

Comment:
Yes to what Alex has suggested, because the `current_assignment` is used to add the definition for the target expression.

This makes me wonder why there's no failure because for something like `x := x + 1` there'll be two definitions. Maybe we might want to add a semantic index test case for named expression in https://github.com/astral-sh/ruff/blob/d37e2e5d33deadc15bf4194216e08ce1528a638c/crates/red_knot_python_semantic/src/semantic_index.rs#L309-L310.

Reply:
Done!

Reply:
Thanks for the suggestion @dhruvmanila ! The plot thickens... this test fails when trying to unwrap the `first_definition`:

```rust
    #[test]
    fn walrus_assignment() {
        let TestCase { db, file } = test_case("x := x + 1");
        let scope = global_scope(&db, file);
        let global_table = symbol_table(&db, scope);

        assert_eq!(names(&global_table), vec!["x"]);

        let use_def = use_def_map(&db, scope);
        let definition = use_def
            .first_public_definition(global_table.symbol_id_by_name("x").unwrap())
            .unwrap();

        assert!(matches!(
            definition.node(&db),
            DefinitionKind::NamedExpression(_)
        ));
    }
```

Some relevant debug info:

```bash
[crates/red_knot_python_semantic/src/semantic_index.rs:494:9] &global_table = SymbolTable {
    symbols: [
        Symbol {
            name: Name("x"),
            flags: SymbolFlags(
                IS_USED,
            ),
        },
    ],
    symbols_by_name: {
        ScopedSymbolId(
            0,
        ): (),
    },
}
[crates/red_knot_python_semantic/src/semantic_index.rs:499:9] &use_def = UseDefMap {
    all_definitions: [],
    all_constraints: [],
    definitions_by_use: [
        SymbolState {
            visible_definitions: Inline(
                [
                    0,
                    0,
                    0,
                ],
            ),
            constraints: [],
            may_be_unbound: true,
        },
        SymbolState {
            visible_definitions: Inline(
                [
                    0,
                    0,
                    0,
                ],
            ),
            constraints: [],
            may_be_unbound: true,
        },
    ],
    public_definitions: [
        SymbolState {
            visible_definitions: Inline(
                [
                    0,
                    0,
                    0,
                ],
            ),
            constraints: [],
            may_be_unbound: true,
        },
    ],
}
[crates/red_knot_python_semantic/src/semantic_index.rs:500:9] &global_table.symbol_id_by_name("x").unwrap() = ScopedSymbolId(
    0,
)
thread 'semantic_index::tests::walrus_assignment' panicked at crates/red_knot_python_semantic/src/semantic_index.rs:503:14:
called `Option::unwrap()` on a `None` value
```

Will try to look into this a little later. (Not sure if it merits a follow-up PR or if it belongs in this one).

Reply:
The issue here is that `x := x + 1` is actually a syntax error, while wrapping it in parens makes it syntactically valid. But our parser is error-recovering, and currently red-knot doesn't issue any diagnostics for invalid syntax, so what ends up happening here is that you just silently don't get any definition of x, because there isn't even an `ExprNamed` in the AST.

This isn't the first time invalid-syntax has caused such confusing symptoms; I think diagnostics on invalid syntax should be a priority for us. I created #13058 to track this.